### PR TITLE
Update clang

### PR DESCRIPTION
**Pull request by filiptronicek**
## Description

We are currently at 14.0.0, but the latest is 17.0.1

### CODE REVIEW

**Review by gengjiawen — APPROVED** [review]

File: chunks/lang-c/Dockerfile [hunk 1]
```diff
@@ -4,7 +4,7 @@ FROM ${base}
 USER root
 
 # Dazzle does not rebuild a layer until one of its lines are changed. Increase this counter to rebuild this layer.
-ENV TRIGGER_REBUILD=1
+ENV TRIGGER_REBUILD=2
 
 RUN curl -fsSL https://apt.llvm.org/llvm-snapshot.gpg.key | sudo gpg --dearmor -o /usr/share/keyrings/llvm-archive-keyring.gpg \
     && echo "deb [signed-by=/usr/share/keyrings/llvm-archive-keyring.gpg] http://apt.llvm.org/jammy/ \
```
Comment (gengjiawen) [inline chunks/lang-c/Dockerfile:10]: following line should be `llvm-toolchain-jammy-17`

**Review by gengjiawen — CHANGES_REQUESTED** [review]

File: chunks/lang-c/Dockerfile (hunk 1 above)
Comment (filiptronicek) [reply chunks/lang-c/Dockerfile:10]: Nice catch, thank you!

**Review by gengjiawen — APPROVED** [review]

**Comment by filiptronicek** [comment]
It seems that we are still using the default version of clang from apt, which is v14. Not sure what's wrong, but will investigate later

**Comment by gengjiawen** [comment]
> It seems that we are still using the default version of clang from apt, which is v14. Not sure what's wrong, but will investigate later

Maybe try official script `yes | bash -c "$(wget -O - https://apt.llvm.org/llvm.sh)"`

**Comment by filiptronicek** [comment]
Thanks for the comments, @gengjiawen. I am not too familiar with C development or this image in particular, so I'll have to dive in deeper when time provides. Marking this as a draft now.

**Comment by gengjiawen** [comment]
> If you have any ideas or findings, please feel free to write them here or make a new PR.

It looks like the package is clang-17, so you should do it like this.

```bash
    && install-packages \
        clang-17 \
        clangd-17 \
        clang-format-17 \
        clang-tidy-17 \
        gdb \
        lld-17
```

Not sure we should make symlink for clang

**Comment by gengjiawen** [comment]
> > If you have any ideas or findings, please feel free to write them here or make a new PR.
> 
> It looks like the package is clang-17, so you should do it like this.
> 
> ```shell
>     && install-packages \
>         clang-17 \
>         clangd-17 \
>         clang-format-17 \
>         clang-tidy-17 \
>         gdb \
>         lld-17
> ```
> 
> Not sure we should make symlink for clang.
> 
> My opinion is to set it as default
> 
> ```
> sudo update-alternatives --install /usr/bin/clang clang /usr/lib/llvm-17/bin/clang 100
> sudo update-alternatives --install /usr/bin/clang++ clang++ /usr/lib/llvm-17/bin/clang++ 100
> ```

@filiptronicek

**Comment by gengjiawen** [comment]
> Thanks for the suggestion, @gengjiawen! It seems that the image builds correctly, but the tests indicate that the executable is not there anymore. I assume the update-alternatives should take care of that, but it doesn't seem like that happens. I can take a look into it next week as well, but if you have any ideas, please do share.
> 
> ```
> failed message="assertion failed: status == 0" result="&test.RunResult{Stderr: []byte("cannot run command: \"exec: \\\"clangd\\\": executable file not found in $PATH\"\nenv:    =/dazzle/runner\n\tPATH=/home/gitpod/.local/bin:/usr/games:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\n\tTRIGGER_REBUILD=2\n\tLANG=en_US.UTF-8\n\tHOME=/home/gitpod\n\tOTEL_TRACES_EXPORTER=otlp\n\tOTEL_EXPORTER_OTLP_TRACES_ENDPOINT=unix:///dev/otel-grpc.sock\n\tOTEL_EXPORTER_OTLP_TRACES_PROTOCOL=grpc\n\tOTEL_TRACE_PARENT=00-7060bb23c7a96ebab31ca62a82f0a958-037b6a6c17a132a1-01\n\tTRACEPARENT=00-7060bb23c7a96ebab31ca62a82f0a958-037b6a6c17a132a1-01\n"), StatusCode: 255}" 
> ```

Please try
```
sudo update-alternatives --install /usr/bin/clangd clangd /usr/bin/clangd-17 100
sudo update-alternatives --install /usr/bin/clang-tidy clang-tidy /usr/bin/clang-tidy-17 100
sudo update-alternatives --install /usr/bin/clang-format clang-format /usr/bin/clang-format-17 100
sudo update-alternatives --install /usr/bin/lld lld /usr/bin/lld-17 100
```

**Comment by filiptronicek** [comment]
Of course, we needed to get alternatives for clangd as well 😄, thanks for noticing! The changes look good now, will merge soon.
image

**Comment by gengjiawen** [comment]
sudo update-alternatives --install /usr/bin/clang clang /usr/lib/llvm-17/bin/clang 100
sudo update-alternatives --install /usr/bin/clang++ clang++ /usr/lib/llvm-17/bin/clang++ 100

Looks missing

**Review by gengjiawen — APPROVED** [review]

**Comment by gengjiawen** [comment]
Let's merge this ?

**Comment by filiptronicek** [comment]
@gengjiawen planning to merge on Monday, Friday didn't quite stick 🤞

**Pull request merged**